crates/ink/tests/return_type_metadata.rs

```diff
@@ -0,0 +1,144 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+#[ink::contract]
+mod contract {
+    #[ink::trait_definition]
+    pub trait TraitDefinition {
+        #[ink(message)]
+        fn get_value(&self) -> u32;
+    }
+
+    #[ink(storage)]
+    pub struct Contract {}
+
+    impl Contract {
+        #[ink(constructor)]
+        pub fn try_new() -> Result<Self, u8> {
+            Err(1)
+        }
+    }
+
+    impl TraitDefinition for Contract {
+        #[ink(message)]
+        fn get_value(&self) -> u32 {
+            42
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use scale_info::{
+        form::PortableForm,
+        Type,
+        TypeDef,
+        TypeDefPrimitive,
+        TypeDefTuple,
+    };
+
+    fn generate_metadata() -> ink_metadata::InkProject {
+        extern "Rust" {
+            fn __ink_generate_metadata() -> ink_metadata::InkProject;
+        }
+
+        unsafe { __ink_generate_metadata() }
+    }
+
+    /// Extract the type defs of the `Ok` and `Error` variants of a `Result` type.
+    ///
+    /// Panics if the type def is not a valid result
+    fn extract_result<'a>(
+        metadata: &'a ink_metadata::InkProject,
+        ty: &'a Type<PortableForm>,
+    ) -> (&'a Type<PortableForm>, &'a Type<PortableForm>) {
+        assert_eq!(
+            "Result",
+            format!("{}", ty.path()),
+            "Message return type should be a Result"
+        );
+        match ty.type_def() {
+            TypeDef::Variant(variant) => {
+                assert_eq!(2, variant.variants().len());
+                let ok_variant = &variant.variants()[0];
+                let ok_field = &ok_variant.fields()[0];
+                let ok_ty = resolve_type(metadata, ok_field.ty().id());
+                assert_eq!("Ok", ok_variant.name());
+
+                let err_variant = &variant.variants()[1];
+                let err_field = &err_variant.fields()[0];
+                let err_ty = resolve_type(metadata, err_field.ty().id());
+                assert_eq!("Err", err_variant.name());
+
+                (ok_ty, err_ty)
+            }
+            td => panic!("Expected a Variant type def enum, got {:?}", td),
+        }
+    }
+
+    /// Resolve a type with the given id from the type registry
+    fn resolve_type(
+        metadata: &ink_metadata::InkProject,
+        type_id: u32,
+    ) -> &Type<PortableForm> {
+        metadata
+            .registry()
+            .resolve(type_id)
+            .unwrap_or_else(|| panic!("No type found in registry with id {}", type_id))
+    }
+
+    #[test]
+    fn trait_message_metadata_return_value_is_result() {
```

Comment:
For the sake of completeness, we may also want to add tests for inherent messages, inherent fallible messages, etc.

Right now as long as the tests in`lang-err-integration-tests` work our metadata stuff is probably fine, but it would be nice to be more explicit with our checks

Reply:
Agree, we need more tests for the metadata like this. Work for a follow-on PR methinks.